solutions/gold/izho-17-HardRoute.mdx

```diff
@@ -0,0 +1,144 @@
+---
+id: izho-17-HardRoute
+source: IZhO 2017
+title: Hard Route
+author: Justin Ji
+---
+
+## Explanation
+
+Assume there is some hard route that goes from vertex $u$ to vertex $v$.
+Let the node that the path from $u$ to $v$ and the furthest node from the hard route meet be node $x$.
+Use rerooting DP to calculate the hardest route and the number of such hardest routes for every node $x$.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N\log{N})$
+
+## Implementation
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+using ll = long long;
+
+int main() {
+	cin.tie(0)->sync_with_stdio(0);
+	int n;
+	cin >> n;
+	vector<vector<int>> adj(n);
+	for (int i = 1; i < n; i++) {
+		int x, y;
+		cin >> x >> y;
+		adj[--x].push_back(--y);
+		adj[y].push_back(x);
+	}
+
+	vector<int> max_length(n), path_count(n);
+	auto dfs = [&](int u, int p, auto &&dfs) -> void {
```

Comment:
the third argument is probably for the recursion i presume
i would suggest just declaring it beforehand with `std::function` since that's the way other sols in the guide do it

Reply:
https://stackoverflow.com/questions/48007520/is-stdfunction-heavier-than-auto-storing-lambda-function
https://codeforces.com/blog/entry/122881

std::function has more overhead, while the type of lambda declaration I wrote shouldn't have much. If you want, I can turn it into a normal function (although global variables are anti-clean, IMO)

Reply:
i mean is your solution that close to the time limit